### ISSUE TITLE
[ENH]: Expanded Figure.axes interface

### ISSUE THREAD

**Issue by timhoffm** [matplotlib/matplotlib#23802]
### Problem

Currently, `Figure.axes` is simply a list of the contained Axes. This is basically what we need for internal data storage, but is not necessarily a great API for getting axes.

Suggestion: I would like to see the possibilty to use `fig.axes[...]` in the same way that I use `fig, axs = plt.subplot*, axs[...]`. e.g.

- `fig.axes[0, 1]` if the Axes are generated via `subplots()`
- `fig.axes['name'] if the Axes are generated via `subplot_mosaic()`
- still also always support the list-like API.

For this we need to make an `AxesCollection` (name t.b.d.) object instead of the simple list.

### Proposed solution

Preliminary idea: Can we simply have an `Axes.id` attribute, which we set to `(0, 1)` in the case of `subplots()`, to 'name' in the case of `subplot_mosaic` and `None` otherwise? Then, `Axes.__getattr__` could simply lookup the `id` for non-integer values. It seems simpler to have this information on the `Axes` objects rather than trying to track it in a separte structure in `AxesCollection`.

**Comment by anntzer** [comment]
You would have to specify what happens if multiple gridspecs or multiple calls to subplot_mosaic (perhaps with the same label names) are used on the same figure.

**Comment by timhoffm** [comment]
> You would have to specify what happens if multiple gridspecs or multiple calls to subplot_mosaic (perhaps with the same label names) are used on the same figure.

I'd simply raise if the given specification is not unique.

Of course this means we have to always run through all axes and check, but these are only a few, so not really an issue.

**Comment by anntzer** [comment]
I am not against considering convenience layers, but I think the one you suggest is not so robust.  For example if I write a library function that takes a figure as argument (which seems like a recommended pattern) and sets up a "foo" axes on it with subplot_mosaic and uses e.g. `figure.axes["foo"]` to access that axes, this will *usually* work, but will fail if the user happens to have added (or later add) another axes with the same name to the figure.  I think we should steer away from this kind of brittleness.

**Comment by jklymak** [comment]
This feels non-unique in lots of other ways.  What would be returned if there are two twinned axes in the same spec?  What about if an axes spans two columns but we only ask for axs[0,0] ?   For a subplot mosaic, do we allow specifying by row and column in addition to name?  

I'm in favour of better tracking where axes came from to better understand what layout goals the user had.  But that seems more than what is proposed here?

**Comment by timhoffm** [comment]
Yes, you can find many ways to make this non-unique. However I estimate that 90%-95% of the time users are only using `subplots()` or `subplot_mosaic()`. I want to support these cases, and I think we can do this if we error out clearly in non-unique cases. So let me re-phrase the proposal:

- Let's add an attribute `Axes.key`; default: None. This `key` can in principle be any object (maybe we want to limit this but that's not fundamental). The `key` does not have to be unique. - *Remark*: this is the same as the proposed `Axes.id` above, but `key` is better because `id` would imply uniqueness, which we don't want to claim.
- `Figure.axes` becomes an `AxesCollection`, which is a wrapper around the axes list. It fulfills the `list` API, and on top allows to index by key following the rule:

  >If there is exactly one Axes in the figure with a given key, `Figure.axes[key]` returns that. Otherwise it errors out.

- `subplots()` sets `key=(i, j)` on the generated Axeses.
- `subplot_mosaic()` sets `key=name` on the generated Axeses.

---

@jklymak in that framework the answers to your questions are:

> What about if an axes spans two columns but we only ask for axs[0,0]

AFAIK you cannot generate spanning `Axes` via `subplots()` - A spanning Axes woud have `key=None` -> (0, 0) is not a valid key and thus errors out.

> For a subplot mosaic, do we allow specifying by row and column in addition to name?

No. Only name.

> What would be returned if there are two twinned axes in the same spec?

We can decide that. I think one can argue both ways:

- the twinned Axes is considered "secondary" -> it does not get a key. -> Indexing by jes returns the original Axes.
- the twinned Axes is considered equivalent -> it gets the same key as the original Axes -> You cannot access via the key anymore:
  > KeyError: There are multiple Axes with key 'my_twin' in the figure

> I'm in favour of better tracking where axes came from to better understand what layout goals the user had. But that seems more than what is proposed here?

Better tracking would be nice. However, I think to be really useful for layout goals, we would need to have a layout structure/definition - e.g. similar to Qt layouts. That's at least an order of magnitude more complex that what I have propsed here.

Here I only want to do something much simpler: We have convenient access API the datastructures `axs` from `fig, axs = plt.subplots/subplot_mosaic`. This information/API is not available in the figure. I thus have to carry around the objects and retrieve them back from the figure.

---
@anntzer I hear your concerns

> but I think the one you suggest is not so robust. 

A general figure can be arbitrary complex. A convenient Axes lookup by key or grid position can thus not work in all cases. The question is: Are we ok to make that work on simple figures and error out when it gets ambiguous? Or do the rare complex cases dictate that we cannot have the convenience for the simple cases?

IMHO this depends on what you promise and how you communicate it. If a user adds an Axes later on, that will also change the current axes and people will live with that.

**Comment by tacaswell** [comment]
I am in favor of this idea, but I think we should put this into accessors on the list rather than over-riding `__getitem__` so something like

```
ax = fig.axes.by_label['bob']
ax = fig.axes.by_index[0, 1]
```

which also means we can support `.items()`, `.keys()`, and `.values()` to round out the `Mapping` interface and we do not have to do type-based dispatch (even though it in unambigious here)

**Comment by timhoffm** [comment]
I may be able to warm to some sort of accessor instead of directly working on `fig.axes`, but need tho think about it a bit more.

**Comment by chahak13** [comment]
Hello! This looks like a nice feature and I am interested in working with you all in implementing this if that's okay?

**Comment by tacaswell** [comment]
I do not think we have reached consensus on what this should look like yet.

A concrete implementation may help to clarify our thinking (see the trade off between different interfaces vs what the implementation would have to look like), but if you do work on this please go knowing that what we ultimately will take may look nothing like your initial work or we may not take it at all.

**Comment by chahak13** [comment]
I understand. I'd like to be a part of the conversation, which is why I decided to work on it. I don't intend to just implement the current suggestion and be done with it. It'll be a good experience working on a feature enhancement with design decisions :)

**Comment by tacaswell** [comment]
@chahak13 Great, you are more than welcome to join this conversation!  I just wanted to make sure expectations are clear up front.

**Comment by jklymak** [comment]
Just to echo that - I think there may only be one proponent for adding this right now, with at least two of us skeptical. That doesn't mean it can't be made into a competing proposal, but its not a ready-to-go idea yet.

**Comment by timhoffm** [comment]
This needs some conceptual work before a (povisional) implementation is reasonable.

The motivation is laid out in the original description: Can we have some id-like mechanism for Axes so that one can easily obtain certain Axes from a figure?

The challenge is to design an the id concept and it's API so that it's intuitive and consistent. This is a bit of a challenge because of the varity of Axes layouts. At least we have:
- names from `subplot_mosaic`
- indices from `subplots`
Additionally we have various other types, which may or may not have reasonable default indices, e.g. twinned, inset, free, spans in a grid etc.

The hard part is really the API, not the implementation. I suppose we'll need a couple of iteration cycles to converge on a good API. During that phase an implementation is not helpful and only poses additional work. The next step should be to write an API proposal (or multiple of these) and illustrate them with code examples.
